src/html/classic/ng/src/web/pages/extras/trashactions.js

```diff
@@ -0,0 +1,181 @@
+        return {restorable: true, deletable: false};
+      },
+      permission: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      port_list: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      report_format: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      role: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      scanner: function() {
+        const restorable = is_defined(entity.credential) ?
+          !entity.credential.isInTrash() : true;
+        return {restorable, deletable: entity.isInUse()};
+      },
+      schedule: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      tag: function() {
+        return {restorable: true, deletable: entity.isInUse()};
+      },
+      target: function() {
+        const ssh_cred = is_defined(entity.ssh_credential) ?
+         !entity.ssh_credential.isInTrash() : true;
+        const smb_cred = is_defined(entity.smb_credential) ?
+         !entity.smb_credential.isInTrash() : true;
+        const esxi_cred = is_defined(entity.esxi_credential) ?
+         !entity.esxi_credential.isInTrash() : true;
+        const snmp_cred = is_defined(entity.snmp_credential) ?
+         !entity.snmp_credential.isInTrash() : true;
+        const portlist = is_defined(entity.port_list) ?
+         !entity.port_list.isInTrash() : true;
+
+        const restorable =
+          ssh_cred && smb_cred && esxi_cred && snmp_cred && portlist;
+        return {restorable, deletable: entity.isInUse()};
+      },
+      task: function() {
+        const restorable = !entity.schedule.isInTrash() &&
+          !entity.target.isInTrash() &&
+          !entity.config.isInTrash() &&
+          !entity.scanner.isInTrash() &&
+          !entity.alerts.some(alert => alert.isInTrash());
+        return {restorable, deletable: false};
+      },
+    };
+
+    function mode() {
+      let restoremode;
+      let deletemode;
+      const can_restore_and_delete = check_by_type[entity.entity_type];
+      const {restorable, deletable} = can_restore_and_delete(entity);
+
+      if (restorable) {
+        restoremode = {
+          title: _('Restore'),
+          onClick: onEntityRestore,
+        };
+      }
+      else {
+        restoremode = {
+          active: false,
+        };
+      }
+
+      if (deletable) {
+        deletemode = {
+          active: false,
+          title: _('Still in use'),
+        };
+      }
+      else {
+        deletemode = {
+          title: _('Delete'),
+          onClick: onEntityDelete,
+        };
+      }
+      return {restoremode, deletemode};
+    }
+
+  return (
+    <TableData>
+      <IconDivider
+        align={['center', 'center']}
+        grow
+      >
+        <RestoreIcon
+          name="restore"
+          value={entity}
+          {...mode().restoremode}
```

Comment:
The mode function is called twice here but you only use one of its returned values. You should either assign its return value to a variable or split it into two different functions one for getting the restoremode and one for getting the deletemode.

Reply:
I would split this into two checks. One for checking if the entity is restoreable and one for checking if the entity is deletable.

Reply:
Also the name mode is a bit confusing because it is about props for the icons. I would expect something like get_delete_props(entity) and get_restore_props(entity) instead of mode. If you still would like to use one function get_delete_restore_props(entity) would be a good name. The api could be
```javascript
const {deleteprops, restorepros} = get_delete_restore_props(entity);
return (
...
  <RestoreIcon
    ...
    {...restoreprops}
  />
...
);
```

Reply:
I renamed the function but sticked to the single function solution. Both checks have to be performed anyway and by keeping both within one function, the source code can be shorter.